Add YACReader

selmf opened the pull request:
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project.
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
    * @luisangelsm (founder) owns and maintains the domain
- [ ] Any additional patches or files have been submitted to the upstream projects concerned.
   * The current qmake build system is deprecated and we are only doing minimal maintenance. Improvements like adding and maintaining an appstream file need to wait until we switch to CMake



[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

hfiguiere commented on com.yacreader.YACReader.appdata.xml:
```diff
@@ -0,0 +1,128 @@
+                </ul>
+            </description>
+        </release>
+    </releases>
```
misising: `content_rating`

hfiguiere commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,123 @@
+  - --socket=wayland
+  - --share=network
+  - --device=dri
+  - --filesystem=home
```
can't the filesystem permissions be narrowed?

selmf replied:
Library management needs directory level access to scan files and create a database. Portals can't provide that and if we limit access to something more restricted like Documents and Downloads only this will be experienced as an unnecessary restriction by the users.
Even worse, during my tests Qt indicated a "driver not loaded" error when the QtSql driver could not access the directories, so lots of bug reports would be expected.

hfiguiere commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,123 @@
+  - --share=ipc
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --share=network
```
does it need network access?

selmf replied:
Yes, the library app has a builtin webserver which can be used to sync comics with a companion iOS app.

hfiguiere commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,123 @@
+          - name: gn
+            buildsystem: simple
+            build-options:
+              env:
+                CC: gcc
+                CXX: g++
```
do you really need to set this?

selmf replied:
Maybe. Depends. The gn bootstrap build system uses environment variables to override default settings. The default in this case is clang. Setting it to gcc saves a build dependency.

hfiguiere replied:
there is a llvm SDK extension if it really want llvm/clang.

hfiguiere commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,123 @@
+        path: com.yacreader.YACReader.appdata.xml
+    modules:
+      - name: unarr
+        buildsystem: cmake
```
```suggestion
        buildsystem: cmake-ninja
```

it's usually faster

hfiguiere commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,123 @@
+      - sub-YACReader-install_subtargets
+      - sub-YACReaderLibrary-install_subtargets
+    post-install:
+      - mv "$FLATPAK_DEST"/share/applications/{YACReader,com.yacreader.YACReader}.desktop
```
you can use `rename-desktop-file` for this

hfiguiere commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,123 @@
+    post-install:
+      - mv "$FLATPAK_DEST"/share/applications/{YACReader,com.yacreader.YACReader}.desktop
+      - mv "$FLATPAK_DEST"/share/applications/{YACReaderLibrary,com.yacreader.YACReader.YACReaderLibrary}.desktop
+      - mv "$FLATPAK_DEST"/share/icons/hicolor/scalable/apps/{YACReader,com.yacreader.YACReader}.svg
```
you can use `rename-icon` for this one.

selmf replied:
There are two desktop files - YACReader.desktop and YACReaderLibrary.desktop. I was able to use rename icon for YACReader, but not for YACReaderLibrary as it gets renamed to YACReader.

hfiguiere commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,123 @@
+      - mv "$FLATPAK_DEST"/share/applications/{YACReaderLibrary,com.yacreader.YACReader.YACReaderLibrary}.desktop
+      - mv "$FLATPAK_DEST"/share/icons/hicolor/scalable/apps/{YACReader,com.yacreader.YACReader}.svg
+      - mv "$FLATPAK_DEST"/share/icons/hicolor/scalable/apps/{YACReaderLibrary,com.yacreader.YACReader.YACReaderLibrary}.svg
+      - sed -i -e 's/^Categories=\(.*\)$/Categories=Office;\1/' -e 's/^Icon=.*$/Icon=com.yacreader.YACReader/' "$FLATPAK_DEST"/share/applications/com.yacreader.YACReader.desktop
+      - sed -i -e 's/^Categories=\(.*\)$/Categories=Office;\1/' -e 's/^Icon=.*$/Icon=com.yacreader.YACReader.YACReaderLibrary/' "$FLATPAK_DEST"/share/applications/com.yacreader.YACReader.YACReaderLibrary.desktop
```
instead of `sed` please use `destkop-file-edit`

Also with the rename icon suggested above, it's taken care of witht the main .desktop file.

selmf replied:
Done - rename for YACReader, desktop-file-edit for YACReaderLibrary.

hfiguiere commented:
bot, build com.yacreader.YACReader

hfiguiere commented:
x86_64 build failure as per https://github.com/flathub/flathub/pull/3522#discussion_r979258044
```
builddir/files/share/appdata/com.yacreader.YACReader.appdata.xml: FAILED:
• tag-missing           : <content_rating> required [use https://odrs.gnome.org/oars]
Validation of files failed
```

hfiguiere commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,123 @@
+        build-commands:
+          # https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=libpdfium-nojs
+          # Fix visibilty for system Freetype
+          - git -C pdfium/build cherry-pick -n bfd6ff0
```
couldn't you just make this a patch to apply and not have to checkout the whole repo?

hfiguiere commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,123 @@
+            build-commands:
+              - ./build/gen.py
+              - ninja -C out
+              - install -D out/gn "$FLATPAK_DEST/bin/gn"
```
```suggestion
              - install -Dm755 out/gn "$FLATPAK_DEST/bin/gn"
```

selmf commented:
I've (hopefully) addressed the issues you spotted and have also updated the build to YACReader 9.9.2 and a newer pdfium module. If we are lucky, this will fix the issues the previous build showed on aarch64.

hfiguiere commented:
bot, build com.yacreader.YACReader

selmf commented:
@hfiguiere It seems the bot is stuck in the clean step of the gn module. Can you cancel it? I need to (re)disable shallow clone for that step. My local cache gave me the false impression it was not needed.

selmf commented:
bot, build com.yacreader.YACReader

selmf commented:
Ok, x64 seems to be working, aarch64 has some troubles. From looking at the logs, this seems to be a toolchain problem. The Pdfium build assumes this is a cross compilation and tries to select a non-existent compiler. I'll write a patch. 

Am 4. Oktober 2022 07:26:13 MESZ schrieb Flathub Bot ***@***.***>:
>Started [test build 113825](https://buildbot.flathub.org/#/builders/32/builds/113825)
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/flathub/flathub/pull/3522#issuecomment-1266413996
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

selmf commented:
bot, build com.yacreader.YACReader

selmf commented:
bot, build com.yacreader.YACReader

barthalion commented on com.yacreader.YACReader.yml:
```diff
@@ -0,0 +1,127 @@
+          - type: git
+            url: https://pdfium.googlesource.com/pdfium
+            # curl 'https://omahaproxy.appspot.com/linux?channel=stable' | cut -d'.' -f 3
+            branch: chromium/5249
```
The problem is precisely this – there's zero guarantee this branch points to the same commit it has been during the build. It may affect reproducibility, should someone want to test whether Flathub build is legit. Wouldn't leaving a comment about which branch was used to find the commit be sufficient?

barthalion commented:
bot, build com.yacreader.YACReader

selmf commented:
@hfiguiere Build works and is ready for review.

barthalion commented:
/merge @luisangelsm
